File: labs OOP/2.3/C++/String.cpp
#include "String.h"
#include <sstream>
#include <cctype>

String::String(const std::string& s) : value(s) {}

std::string String::GetFirstChars() const {
    if (value.empty()) {
        return "";
    }
    std::string result;
    std::istringstream iss(value);
    std::string word;
    while (iss >> word) {
        if (!word.empty()) {
            result += word[0];
        }
    }
    return result;
}

std::string String::GetValue() const {
    return value;
}

void String::UppercaseFirstLetters() {
    if (value.empty()) return;
    bool newWord = true;
    for (char& c : value) {
        if (std::isspace(static_cast<unsigned char>(c))) {
            newWord = true;
        }
        else {
            if (newWord && std::isalpha(static_cast<unsigned char>(c))) {
                c = std::toupper(static_cast<unsigned char>(c));
            }
            newWord = false;
        }
    }
}

File: labs OOP/2.3/C#/String.cs
using System;
using System.Linq;

namespace TextProcessing
{
    public class String : IText
    {
        private string value;

        public String(string s)
        {
            value = s ?? throw new ArgumentNullException(nameof(s));
        }

        public string GetFirstChars()
        {
            if (string.IsNullOrEmpty(value))
            {
                return "";
            }

            return string.Join("", value.Split(' ', StringSplitOptions.RemoveEmptyEntries).Select(word => word[0]));
        }

        public string GetValue()
        {
            return value;
        }

        public void UppercaseFirstLetters()
        {
            if (string.IsNullOrEmpty(value))
            {
                return;
            }

            char[] chars = value.ToCharArray();
            bool newWord = true;

            for (int i = 0; i < chars.Length; i++)
            {
                if (char.IsWhiteSpace(chars[i]))
                {
                    newWord = true;
                }
                else
                {
                    if (newWord && char.IsLetter(chars[i]))
                    {
                        chars[i] = char.ToUpper(chars[i]);
                    }
                    newWord = false;
                }
            }

            value = new string(chars);
        }
    }
}